gradio/helpers.py

```diff
@@ -286,32 +288,85 @@ async def load_example(example_id):
                     outputs=self.outputs,  # type: ignore
                     show_api=False,
                 )
+            if self.cache_examples == "lazy":
+                print(
+                    f"Will cache examples in '{utils.abspath(self.cached_folder)}' directory at first use. ",
+                    end="",
+                )
+                if Path(self.cached_file).exists():
+                    print(
+                        "If method or examples have changed since last caching, delete this folder to reset cache.",
+                        end="",
+                    )
+                print("\n\n")
+                self.cache_logger.setup(self.outputs, self.cached_folder)
+                self.load_input_event.then(
+                    self.lazy_cache,
+                    inputs=[self.dataset],
+                    outputs=self.outputs,
+                    postprocess=False,
+                    api_name=self.api_name,
+                    show_api=False,
+                )
 
-        if self.cache_examples:
+        if self.cache_examples is True:
             if wasm_utils.IS_WASM:
                 # In the Wasm mode, the `threading` module is not supported,
                 # so `client_utils.synchronize_async` is also not available.
                 # And `self.cache()` should be waited for to complete before this method returns,
                 # (otherwise, an error "Cannot cache examples if not in a Blocks context" will be raised anyway)
                 # so `eventloop.create_task(self.cache())` is also not an option.
                 warnings.warn("Caching examples is not supported in the Wasm mode.")
+            elif Path(self.cached_file).exists():
+                print(
+                    f"Using cache from '{utils.abspath(self.cached_folder)}' directory. If method or examples have changed since last caching, delete this folder to clear cache.\n"
+                )
             else:
+                print(f"Caching examples at: '{utils.abspath(self.cached_folder)}'")
+                self.cache_logger.setup(self.outputs, self.cached_folder)
                 client_utils.synchronize_async(self.cache)
 
+    async def _handle_callable_as_generator(self, *args):
+        if self.fn is None:
+            raise ValueError("Cannot lazy-cache examples if no function is provided")
+        if inspect.iscoroutinefunction(self.fn):
+            result = await self.fn(*args)
+            yield result
+        elif inspect.isasyncgenfunction(self.fn):
+            async for item in self.fn(*args):
+                yield item
+        elif inspect.isgeneratorfunction(self.fn):
+            for item in self.fn(*args):
+                yield item
+        else:
+            yield self.fn(*args)
+
+    async def lazy_cache(self, example_index):
+        if Path(self.cached_indices_file).exists():
```

Comment:
The reason I did it this way -- saving the indices of examples that have been cached -- in a separate CSV is because I wanted to preserve one nice feature of cached examples currently, that is that if you cache examples locally and then upload the cached examples to Spaces, you don't need to cache them again. This can save a lot of time for demos that take a long time to run.